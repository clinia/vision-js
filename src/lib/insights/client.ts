import { SearchResults } from '@clinia/search-helper';
import { uniq, createDocumentationMessageGenerator } from '../utils';
import {
  Records,
  InsightsClient,
  InsightsClientMethod,
  InsightsClientPayload,
  InsightsClientWrapper,
  Renderer,
  RendererOptions,
  Unmounter,
  WidgetFactory,
} from '../../types';

const getSelectedRecords = (
  records: Records,
  selectedIDs: string[]
): Records => {
  return selectedIDs.map(id => {
    const record = records.find(h => h.id === id);
    if (typeof record === 'undefined') {
      throw new Error(
        `Could not find id "${id}" passed to \`clickedIdsAfterSearch\` in the returned records. This is necessary to infer the absolute position and the query id.`
      );
    }
    return record;
  });
};

const getQueryID = (selectedRecords: Records): string => {
  const queryIDs = uniq(selectedRecords.map(record => record.__queryID));
  if (queryIDs.length > 1) {
    throw new Error(
      'Insights currently allows a single `queryID`. The `IDs` provided map to multiple `queryID`s.'
    );
  }
  const queryID = queryIDs[0];
  if (typeof queryID !== 'string') {
    throw new Error(
      `Could not infer \`queryID\`. Ensure Vision \`clickAnalytics: true\` was added with the Configure widget.`
    );
  }
  return queryID;
};

const getPositions = (selectedRecords: Records): number[] =>
  selectedRecords.map(record => record.__position);

export const inferPayload = ({
  method,
  results,
  records,
  ids,
}: {
  method: InsightsClientMethod;
  results: SearchResults;
  records: Records;
  ids: string[];
}): Omit<InsightsClientPayload, 'eventName'> => {
  const { index } = results;
  const selectedRecords = getSelectedRecords(records, ids);
  const queryID = getQueryID(selectedRecords);

  switch (method) {
    case 'clickedRecordIDsAfterSearch': {
      const positions = getPositions(selectedRecords);
      return { index, queryID, ids, positions };
    }

    case 'convertedRecordIDsAfterSearch':
      return { index, queryID, ids };

    default:
      throw new Error(`Unsupported method passed to insights: "${method}".`);
  }
};

const wrapInsightsClient = (
  aa: InsightsClient | null,
  results: SearchResults,
  records: Records
): InsightsClientWrapper => (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => {
  if (!aa) {
    const withVisionUsage = createDocumentationMessageGenerator({
      name: 'vision',
    });
    throw new Error(
      withVisionUsage(
        'The `insightsClient` option has not been provided to `vision`.'
      )
    );
  }
  if (!Array.isArray(payload.ids)) {
    throw new TypeError('Expected `ids` to be an array.');
  }
  const inferredPayload = inferPayload({
    method,
    results,
    records,
    ids: payload.ids,
  });
  aa(method, { ...inferredPayload, ...payload } as any);
};

type Connector<TWidgetParams> = (
  renderFn: Renderer<any>,
  unmountFn: Unmounter
) => WidgetFactory<TWidgetParams>;

export default function withInsights(
  connector: Connector<any>
): Connector<unknown> {
  const wrapRenderFn = (
    renderFn: Renderer<RendererOptions<unknown>>
  ): Renderer<RendererOptions<unknown>> => (
    renderOptions: RendererOptions,
    isFirstRender: boolean
  ) => {
    const { results, records, visionInstance } = renderOptions;
    if (results && records && visionInstance) {
      const insights = wrapInsightsClient(
        visionInstance.insightsClient,
        results,
        records
      );
      return renderFn({ ...renderOptions, insights }, isFirstRender);
    }
    return renderFn(renderOptions, isFirstRender);
  };

  return (renderFn: Renderer<RendererOptions<unknown>>, unmountFn: Unmounter) =>
    connector(wrapRenderFn(renderFn), unmountFn);
}
