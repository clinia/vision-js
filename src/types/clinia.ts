import { SearchClient as SearchClientV2 } from 'clinia';
import { SearchResponse as SearchResponseV2 } from '@clinia/client-search';

export type Client = SearchClientV2;

export interface MultiResponse<TRecord = any> {
  results: Array<SearchResponse<TRecord>>;
}

export type SearchResponse<TRecord> = SearchResponseV2<TRecord>;
