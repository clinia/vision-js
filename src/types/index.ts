export * from './clinia';
export * from './connector';
export * from './insights';
export * from './vision';
export * from './widget';
