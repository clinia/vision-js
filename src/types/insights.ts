export type InsightsClientMethod =
  | 'clickedRecordIDsAfterSearch'
  | 'convertedRecordIDsAfterSearch';

export type InsightsClientPayload = {
  eventName: string;
  queryID: string;
  index: string;
  ids: string[];
  positions?: number[];
};

export type InsightsClient = (
  method: InsightsClientMethod,
  payload: InsightsClientPayload
) => void;

export type InsightsClientWrapper = (
  method: InsightsClientMethod,
  payload: Partial<InsightsClientPayload>
) => void;
