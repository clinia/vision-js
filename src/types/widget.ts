import { Index } from '../widgets/index/index';
import {
  CliniaSearchHelper as Helper,
  SearchParameters,
  SearchResults,
  PlainSearchParameters,
} from '@clinia/search-helper';
import { Vision } from './vision';

export interface InitOptions {
  visionInstance: Vision;
  parent: Index | null;
  uiState: UiState;
  state: SearchParameters;
  helper: Helper;
  templatesConfig: object;
  createURL(state: SearchParameters): string;
}

export interface ScopedResult {
  indexId: string;
  results: SearchResults;
  helper: Helper;
}

export interface RenderOptions {
  visionInstance: Vision;
  templatesConfig: object;
  results: SearchResults;
  scopedResults: ScopedResult[];
  state: SearchParameters;
  helper: Helper;
  searchMetadata: {
    isSearchStalled: boolean;
  };
  createURL(state: SearchParameters): string;
}

export interface DisposeOptions {
  helper: Helper;
  state: SearchParameters;
}

export interface WidgetStateOptions {
  searchParameters: SearchParameters;
  helper: Helper;
}

export interface WidgetSearchParametersOptions {
  uiState: IndexUiState;
}

export type IndexUiState = {
  query?: string;
  refinementList?: {
    [property: string]: string[];
  };
  menu?: {
    [property: string]: string;
  };

  toggle?: {
    [property: string]: boolean;
  };
  geoSearch?: {
    /**
     * The rectangular area in geo coordinates.
     * The rectangle is defined by two diagonally opposite points, hence by 4 floats separated by commas.
     *
     * @example '47.3165,4.9665,47.3424,5.0201'
     */
    boundingBox: string;
  };
  page?: number;
  perPage?: number;
  configure?: PlainSearchParameters;
  places?: {
    query: string;
    /**
     * The central geolocation.
     *
     * @example '48.8546,2.3477'
     */
    position: string;
  };
};

export type UiState = {
  [indexId: string]: IndexUiState;
};

/**
 * Widgets are the building blocks of Visionjs. Any valid widget must
 * have at least a `render` or a `init` function.
 */
export interface Widget {
  $$type?:
    | 'cvi.autocomplete'
    | 'cvi.breadcrumb'
    | 'cvi.clearRefinements'
    | 'cvi.configure'
    | 'cvi.configureRelatedItems'
    | 'cvi.currentRefinements'
    | 'cvi.geoSearch'
    | 'cvi.hierarchicalMenu'
    | 'cvi.records'
    | 'cvi.perPage'
    | 'cvi.index'
    | 'cvi.infiniteRecords'
    | 'cvi.menu'
    | 'cvi.numericMenu'
    | 'cvi.pagination'
    | 'cvi.places'
    | 'cvi.poweredBy'
    | 'cvi.queryRules'
    | 'cvi.queryRuleCustomData'
    | 'cvi.queryRuleContext'
    | 'cvi.range'
    | 'cvi.rangeInput'
    | 'cvi.rangeSlider'
    | 'cvi.ratingMenu'
    | 'cvi.refinementList'
    | 'cvi.searchBox'
    | 'cvi.sortBy'
    | 'cvi.stats'
    | 'cvi.toggleRefinement'
    | 'cvi.voiceSearch';
  /**
   * Called once before the first search
   */
  init?(options: InitOptions): void;
  /**
   * Called after each search response has been received
   */
  render?(options: RenderOptions): void;
  /**
   * Called when this widget is unmounted. Used to remove refinements set by
   * during this widget's initialization and life time.
   */
  dispose?(options: DisposeOptions): SearchParameters | void;
  /**
   * This function is required for a widget to be taken in account for routing.
   * It will derive a uiState for this widget based on the existing uiState and
   * the search parameters applied.
   * @param uiState current state
   * @param widgetStateOptions extra information to calculate uiState
   */
  getWidgetState?(
    uiState: IndexUiState,
    widgetStateOptions: WidgetStateOptions
  ): IndexUiState;
  /**
   * This function is required for a widget to behave correctly when a URL is
   * loaded via e.g. routing. It receives the current UiState and applied search
   * parameters, and is expected to return a new search parameters.
   * @param state applied search parameters
   * @param widgetSearchParametersOptions extra information to calculate next searchParameters
   */
  getWidgetSearchParameters?(
    state: SearchParameters,
    widgetSearchParametersOptions: WidgetSearchParametersOptions
  ): SearchParameters;
}

export type WidgetFactory<TWidgetParams> = (
  widgetParams: TWidgetParams
) => Widget;

export type Template<TTemplateData = void> =
  | string
  | ((data: TTemplateData) => string);
