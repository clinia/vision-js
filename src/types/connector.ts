import { SearchResults } from '@clinia/search-helper';
import { Records, Vision } from './vision';
import { InsightsClient } from './insights';

export type RendererOptions<TWidgetParams = unknown> = {
  /**
   * Original parameters for this widget.
   * Useful for giving back the render parameters to the renderer.
   */
  widgetParams: TWidgetParams;
  visionInstance: Vision;
  results?: SearchResults;
  records?: Records;
  insights?: InsightsClient;
};

export type Renderer<TRenderOptions extends RendererOptions = any> = (
  renderOptions: TRenderOptions,
  isFirstRender: boolean
) => void;

export type Unmounter = () => void;
