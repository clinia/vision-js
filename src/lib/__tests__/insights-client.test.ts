import { withInsights, inferInsightsPayload } from '../insights';
import { Widget, WidgetFactory } from '../../types';

const connectRecords = (
  renderFn: any,
  unmountFn: any
): WidgetFactory<unknown> => (widgetParams = {}): Widget => ({
  init() {},
  render({ results, visionInstance }) {
    const records = results.records;
    renderFn({ records, results, visionInstance, widgetParams }, false);
  },
  dispose() {
    unmountFn();
  },
});

const createWidgetWithInsights = ({
  renderFn,
  visionInstance,
  results,
}): Widget => {
  const connectRecordsWithInsights = withInsights(connectRecords as any);
  const widget = connectRecordsWithInsights(renderFn, jest.fn())({});
  (widget as any).render({ results, visionInstance } as any);
  return widget;
};

describe('withInsights', () => {
  describe('when applied on connectRecords', () => {
    it('should call the passed renderFn', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      expect(renderFn).toHaveBeenCalledTimes(1);
    });

    it('should not remove any renderProps passed by connectRecords', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(renderProps).toEqual(
        expect.objectContaining({
          visionInstance,
          results,
          widgetParams: {},
        })
      );
    });

    it('should expose the insights client wrapper to renderOptions if passed to visionInstance', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(renderProps).toHaveProperty('insights');
      expect(renderProps.insights).toBeInstanceOf(Function);
    });
    it('should expose the insights client wrapper even when insightsClient was not provided', () => {
      const renderFn = jest.fn();
      const visionInstance = {
        /* insightsClient was not passed */
      };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(renderProps).toHaveProperty('insights');
      expect(renderProps.insights).toBeInstanceOf(Function);
    });
    it('should expose the insights client wrapper that throws when insightsClient was not provided', () => {
      const renderFn = jest.fn();
      const visionInstance = {
        /* insightsClient was not passed */
      };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(() => {
        renderProps.insights('clickedRecordIDsAfterSearch', {
          eventName: 'add to favorites',
          ids: ['1'],
        });
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('exposed insights client wrapper', () => {
    it('should call the insights client under the hood', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      renderProps.insights('clickedRecordIDsAfterSearch', {
        ids: ['3'],
        eventName: 'Add to basket',
      });
      expect(visionInstance.insightsClient).toHaveBeenCalledTimes(1);
    });

    it('should pass it the correct parameters', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      renderProps.insights('clickedRecordIDsAfterSearch', {
        ids: ['3'],
        eventName: 'Add to basket',
      });
      const [method, payload] = visionInstance.insightsClient.mock.calls[0];
      expect(method).toEqual('clickedRecordIDsAfterSearch');
      expect(payload).toEqual({
        eventName: 'Add to basket',
        index: 'theIndex',
        queryID: 'theQueryID',
        ids: ['3'],
        positions: [11],
      });
    });

    it('should not infer or pass the positions if method is `convertedRecordIDsAfterSearch`', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      renderProps.insights('convertedRecordIDsAfterSearch', {
        ids: ['1'],
        eventName: 'Add to basket',
      });
      const [method, payload] = visionInstance.insightsClient.mock.calls[0];
      expect(method).toEqual('convertedRecordIDsAfterSearch');
      expect(payload).toEqual({
        eventName: 'Add to basket',
        index: 'theIndex',
        queryID: 'theQueryID',
        ids: ['1'],
      });
    });

    it('should reject non-existing ids', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(() => {
        renderProps.insights('clickedRecordIDsAfterSearch', {
          ids: ['xxxxxx'],
          eventName: 'Add to basket',
        });
      }).toThrowErrorMatchingInlineSnapshot(
        `"Could not find id \\"xxxxxx\\" passed to \`clickedIdsAfterSearch\` in the returned records. This is necessary to infer the absolute position and the query id."`
      );
    });

    it('should reject if ids provided have different queryIDs', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __position: 1, __queryID: 'theQueryID_1' },
          { id: '2', __position: 2, __queryID: 'theQueryID_2' },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(() => {
        renderProps.insights('clickedRecordIDsAfterSearch', {
          ids: ['1', '2'],
          eventName: 'Add to basket',
        });
      }).toThrowErrorMatchingInlineSnapshot(
        `"Insights currently allows a single \`queryID\`. The \`IDs\` provided map to multiple \`queryID\`s."`
      );
    });

    it('should reject if no queryID found (clickAnalytics was not set to true)', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __position: 1 },
          { id: '2', __position: 2 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(() => {
        renderProps.insights('clickedRecordIDsAfterSearch', {
          ids: ['1', '2'],
          eventName: 'Add to basket',
        });
      }).toThrowErrorMatchingInlineSnapshot(
        `"Could not infer \`queryID\`. Ensure Vision \`clickAnalytics: true\` was added with the Configure widget."`
      );
    });

    it('should reject unknown method name', () => {
      const renderFn = jest.fn();
      const visionInstance = { insightsClient: jest.fn() };
      const results = {
        index: 'theIndex',
        records: [
          { id: '1', __queryID: 'theQueryID', __position: 9 },
          { id: '2', __queryID: 'theQueryID', __position: 10 },
          { id: '3', __queryID: 'theQueryID', __position: 11 },
          { id: '4', __queryID: 'theQueryID', __position: 12 },
        ],
      };
      createWidgetWithInsights({ renderFn, visionInstance, results });

      const [renderProps] = renderFn.mock.calls[0];
      expect(() => {
        renderProps.insights('unknow_method', {
          ids: ['3'],
          eventName: 'Add to basket',
        });
      }).toThrowErrorMatchingInlineSnapshot(
        `"Unsupported method passed to insights: \\"unknow_method\\"."`
      );
    });
  });
});

describe('inferInsightsPayload', () => {
  const records: any = [
    { id: '1', __queryID: 'theQueryID', __position: 9 },
    { id: '2', __queryID: 'theQueryID', __position: 10 },
    { id: '3', __queryID: 'theQueryID', __position: 11 },
    { id: '4', __queryID: 'theQueryID', __position: 12 },
  ];
  const results: any = {
    index: 'theIndex',
    records,
  };

  describe('payload inferring', () => {
    it('should infer queryID from results', () => {
      const payload = inferInsightsPayload({
        method: 'clickedRecordIDsAfterSearch',
        results,
        records,
        ids: ['3'],
      });
      expect(payload.queryID).toEqual('theQueryID');
    });

    it('should infer index name from results', () => {
      const payload = inferInsightsPayload({
        method: 'clickedRecordIDsAfterSearch',
        results,
        records,
        ids: ['3'],
      });
      expect(payload.index).toEqual(results.index);
    });

    it('should inject passed ids', () => {
      const payload = inferInsightsPayload({
        method: 'clickedRecordIDsAfterSearch',
        results,
        records,
        ids: ['3', '4'],
      });
      expect(payload.ids).toEqual(['3', '4']);
    });

    it('should compute and inject hit positions', () => {
      const payload = inferInsightsPayload({
        method: 'clickedRecordIDsAfterSearch',
        results,
        records,
        ids: ['3', '4'],
      });
      expect(payload.positions).toEqual([11, 12]);
    });
  });
});
