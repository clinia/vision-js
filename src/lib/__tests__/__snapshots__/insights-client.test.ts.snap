// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withInsights when applied on connectRecords should expose the insights client wrapper that throws when insightsClient was not provided 1`] = `
"The \`insightsClient\` option has not been provided to \`vision\`.

See documentation: https://www.clinia.com/doc/api-reference/widgets/vision/js/"
`;
